### PR TITLE
Enable the `no-unused-vars` ESLint rule

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Please see http://eslint.org/docs/rules/no-unused-vars; note that this patch purposely uses the same rule options as in `mozilla-central`, such that it fixes part of issue #7957.

It wasn't, in my opinion, entirely straightforward to enable this rule compared to the already existing rules. In many cases a `var descriptiveName = ...` format was used (more or less) to document the code, and I choose to place the old variable name in a trailing comment to not lose that information.
~~In other cases I simply commented out the "unused" lines instead of outright removing them, since that's already done in some places in the code-base to preserve the logic for future use. (Since comments are stripped by the preprocessor anyway, that shouldn't be a problem IMHO.)~~

I welcome feedback on these changes, since it wasn't always entirely easy to know what changes made the most sense in every situation.

### CODE REVIEW

**Comment by timvandermeij** [comment]
The trailing comments solution makes sense, but in my opinion commenting out code should be avoided at all costs. I have seen it in other projects before and for me it adds nothing but confusion about why it's commented out. Is the code buggy? Is it intended for future use? It is just unused? I think that for all these questions the answer should be the same: remove the code; we have version control for that. Buggy or unused code just adds noise and makes the code less readable. Preserving logic for future use should not be done in the main tree. It should be developed on a branch and only merged if it's actually used. Moreover, removed code can always be recovered because of version control, in the rare case that one would want to use it later on.

Personally I'd rather not enable this rule and have cleaner code (only removing the "really" unused variables like https://github.com/mozilla/pdf.js/pull/7972/files#diff-18ca06e2bd5be4c9132b51f78388b8f0L41) than enable this rule and have to deal with commented out code, but again, that's my opinion and those of others may differ.

**Comment by Snuffleupagus** [comment]
Unfortunately we've got "prior art" when it comes to keeping commented-out code around, note in particular [src/core/evaluator.js#L1326-L1336](https://github.com/mozilla/pdf.js/blob/f8e793f8c2d5de791fea52b234a51e3421e070c6/src/core/evaluator.js#L1326-L1336), which is why I figured it might be OK in moderation (that existing code being commented actually caused all of `evaluator.js` changes made in this patch).

**Comment by timvandermeij** [comment]
You're right, and that is exactly why I don't think we should add more. I have no idea what to make of that commented out piece of code. Why is it commented out? Will it ever be used again? It just results in more questions than answers, really.

However, I do see why you chose this approach for the patch, given the fact that commented out is already present in the codebase. Therefore, please note that this is purely my opinion on the matter and not a critique towards this specific patch!

**Comment by Snuffleupagus** [comment]
> Therefore, please note that this is purely my opinion on the matter and not critique towards this specific patch!

I'm not exactly happy about the current situation either; so no worries, I understand what you meant here :-)

---

@yurydelendik Given that `no-unused-vars` is one of the rules you specifically mentioned in https://github.com/mozilla/pdf.js/issues/7957#issuecomment-272467569, how do you think that we should proceed here?
Can we add even more commented-out code, or should we just remove those lines instead?

**Comment by yurydelendik** [comment]
🤔

**Comment by Snuffleupagus** [comment]
In case it helps: The latest version of the PR only comments out code in `evaluator.js`, which was necessitated by existing commented-out code, please see [evaluator.js#L1326-L1336](https://github.com/mozilla/pdf.js/blob/e0a92a7f4853adb4145145920f2c85078d1695ff/src/core/evaluator.js#L1326-L1336).

That code was added in PR #4570 and it apparently *landed* in its current (commented-out) state, meaning that this code has never actually been used; see https://github.com/mozilla/pdf.js/pull/4570/files#diff-0b94c2e77a5259f7a728122fdbf9f46aR785.

**Comment by brendandahl** [comment]
I'd say remove commented out code, unless it's in a jsdoc as an example.  As for https://github.com/mozilla/pdf.js/blob/e0a92a7f4853adb4145145920f2c85078d1695ff/src/core/evaluator.js#L1326 we should move that to an issue for when it's needed (we may already have one?).

**Comment by Snuffleupagus** [comment]
> I'd say remove commented out code, unless it's in a jsdoc as an example. As for https://github.com/mozilla/pdf.js/blob/e0a92a7f4853adb4145145920f2c85078d1695ff/src/core/evaluator.js#L1326 we should move that to an issue for when it's needed (we may already have one?).

@brendandahl Thanks so much for the feedback here!
I've removed that code in this PR, and opened issue #7996 for tracking potentially using the code again.

File: test/driver.js [hunk 1]
```diff
@@ -673,3 +673,5 @@ var Driver = (function DriverClosure() {
 
   return Driver;
 })();
+
+exports.Driver = Driver;
```
Comment (timvandermeij) [inline test/driver.js]: I don't really understand this change. There is no `exports` variable in this file. Could you explain this a bit?

File: test/font/fontutils.js [hunk 2]
```diff
@@ -88,3 +88,7 @@ function verifyTtxOutput(output) {
     throw m[1];
   }
 }
+
+exports.decodeFontData = decodeFontData;
```
Comment (timvandermeij) [inline test/font/fontutils.js]: I don't really understand these changes. There is no `exports` variable in this file. Could you explain this a bit?

File: test/unit/testreporter.js [hunk 3]
```diff
@@ -74,3 +74,5 @@ var TestReporter = function(browser, appPath) {
     setTimeout(sendQuitRequest, 500);
   };
 };
+
+exports.TestReporter = TestReporter;
```
Comment (timvandermeij) [inline test/unit/testreporter.js]: I don't really understand this change. There is no `exports` variable in this file. Could you explain this a bit?

File: test/unit/network_spec.js [hunk 4]
```diff
@@ -94,7 +94,7 @@ describe('network', function() {
         rangeChunkSize: 65536,
         disableStream: false,
       },
-      disableRange: false
+      disableRange: true
```
Comment (timvandermeij) [inline test/unit/network_spec.js]: Why is this change made? It looks unrelated to the patch. The same goes for the other changes in this file. Could you explain this a bit?

**Review by timvandermeij — COMMENTED** [review]
Looks good overall, but I need some more information before approving this. Thanks.

File: test/driver.js (hunk 1 above)
Comment (Snuffleupagus) [inline test/driver.js]: Without any changes to this file, the `no-unused-vars` rule would complain since `Driver` isn't called from within this file (and this seemed better than adding a bunch more `// eslint-disable-line no-unused-vars`).

Please keep in mind here that this file, as with most/all test files, are run using Node.js where `exports` is defined. Note how https://github.com/mozilla/pdf.js/blob/master/test/.eslintrc lists `node` as one of the environments, and test files such as e.g. https://github.com/mozilla/pdf.js/blob/master/test/downloadutils.js#L170-L171 already uses `exports`.

I really ought to have mentioned this since it's not that obvious, but hopefully the above clears it up!

File: test/font/fontutils.js (hunk 2 above)
Comment (Snuffleupagus) [inline test/font/fontutils.js]: Please see https://github.com/mozilla/pdf.js/pull/7972#discussion_r98364616.

File: test/unit/testreporter.js (hunk 3 above)
Comment (Snuffleupagus) [inline test/unit/testreporter.js]: Please see https://github.com/mozilla/pdf.js/pull/7972#discussion_r98364616.

File: test/unit/network_spec.js (hunk 4 above)
Comment (Snuffleupagus) [inline test/unit/network_spec.js]: Well, I suppose that you're correct about this being unrelated!
Since I needed to make a couple of adjustments in this file anyway, I figured that it couldn't hurt to clean up a couple of things. But your comment makes me realize that I may have gone overboard here, so I'll revert this to the essential changes only. Thank you!

Comment (Snuffleupagus) [inline test/unit/network_spec.js]: Please note that in the updated patch, there's no functional changes.
However, in order to keep the same general structure of all the tests, I've added some currently missing `expect` statements, and the same Promise handling throughout, rather than removing code. I hope this is deemed acceptable!

**Comment by timvandermeij** [comment]
Looks good. Thank you for working on this!
